agent/hcp/telemetry/exporter_test.go

```diff
@@ -0,0 +1,245 @@
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/instrumentation"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+func TestExporter_NewMetricsExporter(t *testing.T) {
+	for name, tc := range map[string]struct {
+		cfg         *MetricsExporterConfig
+		wantErr     string
+		expectedExp *MetricsExporter
+	}{
+		"failsWithoutClient": {
+			cfg: &MetricsExporterConfig{
+				Client:  nil,
+				Filters: []string{"raft.apply$"},
+				Logger:  hclog.New(&hclog.LoggerOptions{Output: io.Discard}),
+			},
+			wantErr: "HCP client and a logger are required",
+		},
+		"failsWithoutLogger": {
+			cfg: &MetricsExporterConfig{
+				Client:  nil,
+				Filters: []string{"raft.apply$"},
+			},
+			wantErr: "HCP client and a logger are required",
+		},
+		"invalidFilter": {
+			cfg: &MetricsExporterConfig{
+				Client: hcpclient.NewMockClient(t),
+				// Unsupported re2 regex syntax
+				Filters: []string{"(*LF)"},
+				Logger:  hclog.New(&hclog.LoggerOptions{Output: io.Discard}),
+			},
+			wantErr: "invalid regex",
+		},
+		"success": {
+			cfg: &MetricsExporterConfig{
+				Client:  hcpclient.NewMockClient(t),
+				Filters: []string{"raft.apply$"},
+				Labels: map[string]string{
+					"instance.id": "asdkfj",
+				},
+				Logger: hclog.New(&hclog.LoggerOptions{Output: io.Discard}),
+			},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			exp, err := NewMetricsExporter(tc.cfg)
+			if tc.wantErr != "" {
+				require.Error(t, err)
+				require.Contains(t, err.Error(), tc.wantErr)
+				require.Nil(t, exp)
+			} else {
+				for _, v := range tc.cfg.Filters {
+					_, ok := exp.filter.filters[v]
+					require.True(t, ok)
+				}
+				attr := make([]attribute.KeyValue, 0, len(tc.cfg.Labels))
+				for k, v := range tc.cfg.Labels {
+					attr = append(attr, attribute.KeyValue{
+						Key:   attribute.Key(k),
+						Value: attribute.StringValue(v),
+					})
+				}
+				require.Equal(t, exp.resource, resource.NewWithAttributes("", attr...))
+				require.NotNil(t, exp.logger)
+				require.NotNil(t, exp.client)
+			}
+		})
+	}
+}
+
+func TestExporter_Export(t *testing.T) {
+	ctx := context.Background()
+
+	client := hcpclient.NewMockClient(t)
+	cfg := &MetricsExporterConfig{
+		Client:  client,
+		Filters: []string{"raft.*"},
+		Labels: map[string]string{
+			"instance.id": "testserver",
+		},
+		Logger: hclog.New(&hclog.LoggerOptions{Output: io.Discard}),
+	}
+	exp, err := NewMetricsExporter(cfg)
+	require.NoError(t, err)
+
+	now := time.Now()
+	later := now.Add(5 * time.Second)
+	goMetrics := []*metrics.IntervalMetrics{
+		{
+			Interval: now,
+			Gauges: map[string]metrics.GaugeValue{
+				"consul.raft.peers": {
+					Name:  "consul.raft.peers",
+					Value: 2.0,
+					Labels: []metrics.Label{
+						{
+							Name:  "peer_type",
+							Value: "test",
+						},
+					},
+				},
+				"filtered_gauges": {
```

Comment:
TODO: should probably refactor out this filter case :p tried to do it all in one test case...